Container query with font units should invalidate when font changes

anttijk opened the pull request:
#### 0b06dbb375788621d903a2be904c2cf595557d85
<pre>
Container query with font units should invalidate when font changes
<a href="https://bugs.webkit.org/show_bug.cgi?id=253940">https://bugs.webkit.org/show_bug.cgi?id=253940</a>
rdar://106739736

Reviewed by Tim Nguyen.

* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/container-queries/font-relative-calc-dynamic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/container-queries/font-relative-units-dynamic-expected.txt:

The remaining failures are units we don&apos;t support.

* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::styleChangeAffectsRelativeUnits):

Factor into a function.

(WebCore::Style::TreeResolver::resolveElement):
(WebCore::Style::TreeResolver::updateStateForQueryContainer):

If the container font size changes we need to evaluate all descendants in case there are container queries that use
font-relative units.

* Source/WebCore/style/StyleTreeResolver.h:

Canonical link: <a href="https://commits.webkit.org/267258@main">https://commits.webkit.org/267258@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/10138147185565112f6b082f9e565aab8f1d120a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16125 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16442 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/16847 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/17887 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15130 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16310 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/19473 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16550 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/17543 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16318 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/16785 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/13771 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18651 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14028 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14589 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/21428 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15017 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/14754 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/17985 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15345 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/13010 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14572 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3843 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/18941 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15166 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -937,7 +937,7 @@ const RenderStyle* TreeResolver::existingStyle(const Element& element)
 }
 
-auto TreeResolver::updateStateForQueryContainer(Element& element, const RenderStyle* style, ContainerType previousContainerType, Change& change, DescendantsToResolve& descendantsToResolve) -> QueryContainerAction
+auto TreeResolver::updateStateForQueryContainer(Element& element, const RenderStyle* style, const RenderStyle* existingStyle, Change& change, DescendantsToResolve& descendantsToResolve) -> QueryContainerAction
```
nit: usually we put old before new, but I don't feel too strongly

anttijk replied:
who is "we" in this?

nt1m replied:
I often see `oldStyle` before `newStyle` in the codebase

nt1m commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -955,7 +955,10 @@ auto TreeResolver::updateStateForQueryContainer(Element& element, const RenderSt
+    if (style->containerType() != ContainerType::Normal || (existingStyle && existingStyle->containerType() != ContainerType::Normal)) {
+        // If any of the queries use font-size relative units then a font size change may affect their evaluation.
+        if (existingStyle && style->computedFontSize() != existingStyle->computedFontSize())
+            descendantsToResolve = DescendantsToResolve::All;
```
Would it be useful to store whether a container query has font relative units to avoid doing unnecessary work if the container query doesn't have any?

anttijk replied:
That's a possible optimization but in practice the case is rare. Also normally changing an inherited property (like font properties) will cause all descendants to be recomputed in any case. The test case specifically stops this.

anttijk committed:
Container query with font units should invalidate when font changes
https://bugs.webkit.org/show_bug.cgi?id=253940
rdar://106739736

Reviewed by Tim Nguyen.

* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/container-queries/font-relative-calc-dynamic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/container-queries/font-relative-units-dynamic-expected.txt:

The remaining failures are units we don't support.

* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::styleChangeAffectsRelativeUnits):

Factor into a function.

(WebCore::Style::TreeResolver::resolveElement):
(WebCore::Style::TreeResolver::updateStateForQueryContainer):

If the container font size changes we need to evaluate all descendants in case there are container queries that use
font-relative units.

* Source/WebCore/style/StyleTreeResolver.h:

Canonical link: https://commits.webkit.org/267258@main
